augur/dates/__init__.py

```diff
@@ -200,39 +200,17 @@ def is_date_ambiguous(date, ambiguous_by="any"):
         "X" in day and ambiguous_by in ("any", "day")
     ))
 
-def get_numerical_date_from_value(value, fmt=None, min_max_year=None):
-    value = str(value)
-    if re.match(r'^-*\d+\.\d+$', value):
-        # numeric date which can be negative
-        return float(value)
```

Comment:
Hmm. I see now we were gating here with a regex too. The context I'm missing at the moment is which code path was used where and when (because I haven't traced it), and thus where only this specific float literal syntax was supported.

Reply:
Code paths are important to note. My impression is that date parsing across Augur has come about in the form of convergent evolution, with separate functions trying to do roughly the same thing. These are the entry points I've gathered for what's currently on `master`:

- `numeric_date()`
    - is used by:
        - `--min-date/--max-date` options to `augur filter` and `augur frequencies` (via `numeric_date_type()`)
        - some example snippets in docstrings
- `get_numerical_date_from_value()`
    - is used by:
        - parsing of the metadata `date` column in `augur filter`, `augur frequencies`, `augur refine` (via `get_numerical_dates()`)
        - parsing of the metadata `date` column in `augur parse`
- `is_date_ambiguous()`
    - is used by `augur filter --exclude-ambiguous-dates-by`
    - can only handle `<year>-<month>-<day>`, `<year>-<month>`, or `<year>` dates
- `get_iso_year_week()`
    - is used by `augur filter` when called with `--group-by week`

Also note there is some date parsing that currently lives outside of the `dates` module:

- `augur filter` when called with `--group-by <year|month|week>` parses dates in the form of `<year>-<month>-<day>`, `<year>-<month>`, or `<year>` ([src](https://github.com/nextstrain/augur/blob/e7965fcb81c1a47d4c0c8d82e6e1641879ca5200/augur/filter/subsample.py#L140-L151))

It feels like a mess (which I definitely contributed to), and I'd like to clean it up a bit hence my recent PRs.

Reply:
> It feels like a mess … and I'd like to clean it up a bit hence my recent PRs.

:raised_hands: Agreed, so +1 for that. :-)